File: Threads Posix/exo4.py
import threading
import random
import time
import sys
import os

class Tableau:
    def __init__(self, n):
        self.tab = []
        self.n = n

class Paquet:
    def __init__(self, t1, t2):
        self.t1 = t1
        self.t2 = t2

def generer(tableau):
    tableau.tab = [random.randint(0, 99) for _ in range(tableau.n)]

def verifier_inclusion(paquet):
    inclus = True
    for x in paquet.t1.tab:
        if x not in paquet.t2.tab:
            inclus = False
            break
    
    resultat = "inclus" if inclus else "NON inclus"
    print(f"\nRésultat : T1 est {resultat} dans T2")

def guetter_annulation():
    if sys.stdin.read(1) == 'A':
        print("\nProgramme annulé par l'utilisateur.")
        os._exit(0)

def main():
    random.seed(time.time())
    
    t1 = Tableau(5)
    t2 = Tableau(10)
    p = Paquet(t1, t2)

    th_abort = threading.Thread(target=guetter_annulation, daemon=True)
    th_abort.start()

    th1 = threading.Thread(target=generer, args=(t1,))
    th2 = threading.Thread(target=generer, args=(t2,))

    th1.start()
    th2.start()

    th1.join()
    th2.join()

    th3 = threading.Thread(target=verifier_inclusion, args=(p,))
    th3.start()
    th3.join()

if __name__ == "__main__":
    main()

File: Threads Posix/exo4.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <time.h>

typedef struct Tableau
{ 
    int *tab; 
    int n; 
} Tableau;
typedef struct Paquet
{ 
    Tableau *T1; 
    Tableau *T2; 
} Paquet;


void* generer(void* arg) 
{
    Tableau *t = (Tableau*)arg;
    t->tab = malloc(t->n * sizeof(int));
    for (int i = 0; i < t->n; i++) 
    {
        t->tab[i] = rand() % 100;
    }
    return NULL;
}


void* verifier_inclusion(void* arg) 
{
    Paquet *p = (Paquet*)arg;
    int inclus = 1;
    for (int i = 0; i < p->T1->n; i++) 
    {
        int trouve = 0;
        for (int j = 0; j < p->T2->n; j++) 
        {
            if (p->T1->tab[i] == p->T2->tab[j]) 
            { 
                trouve = 1; 
                break; 
            }
        }
        if (!trouve) 
        {
            inclus = 0; 
            break; 
        }
    }
    printf("\nRésultat : T1 est %s dans T2\n", inclus ? "inclus" : "NON inclus");
    return NULL;
}


void* guetter_annulation(void* arg) 
{
    if (getchar() == 'A') 
    {
        printf("\nProgramme annulé par l'utilisateur.\n");
        exit(0);
    }
    return NULL;
}

int main() 
{
    srand(time(NULL));
    pthread_t th1, th2, th3, th_abort;
    Tableau t1 = {NULL, 5}, t2 = {NULL, 10};
    Paquet p = {&t1, &t2};

    
    pthread_create(&th_abort, NULL, guetter_annulation, NULL);
    pthread_detach(th_abort); 

    pthread_create(&th1, NULL, generer, &t1);
    pthread_create(&th2, NULL, generer, &t2);

    pthread_join(th1, NULL);
    pthread_join(th2, NULL);

    
    pthread_create(&th3, NULL, verifier_inclusion, &p);
    pthread_join(th3, NULL);

    free(t1.tab); free(t2.tab);
    return 0;
}